File: sorting/bubblesort.py
"""
Bubble Sort Algorithm Implementation in Python

Time Complexity:
- Best Case: O(n) - when array is already sorted (optimized version)
- Average Case: O(n²)
- Worst Case: O(n²) - when array is reverse sorted
Space Complexity: O(1) for in-place, O(n) for functional approach

Bubble Sort works by repeatedly stepping through the list, comparing adjacent
elements and swapping them if they are in the wrong order. The pass through
the list is repeated until the list is sorted.

Python features:
- Type hints for clarity
- Multiple implementation patterns (recursive, iterative, optimized)
- Generator expressions
- List comprehensions
- Object-oriented design
"""

from typing import List, TypeVar, Callable, Tuple
import time
import random
from copy import deepcopy

T = TypeVar("T")


def bubble_sort_iterative(arr: List[T]) -> List[T]:
    """
    Basic iterative bubble sort implementation.

    This is the standard bubble sort algorithm that compares and swaps
    adjacent elements until the entire array is sorted.

    Args:
        arr: List to be sorted

    Returns:
        New sorted list

    Time Complexity: O(n²) in all cases (no optimization)
    Space Complexity: O(n) for the new list
    """
    if len(arr) <= 1:
        return arr.copy()

    result = arr.copy()
    n = len(result)

    # Outer loop for number of passes
    for i in range(n):
        # Inner loop for comparisons
        # After each pass, the largest element "bubbles up" to its position
        for j in range(0, n - i - 1):
            if result[j] > result[j + 1]:
                # Swap adjacent elements
                result[j], result[j + 1] = result[j + 1], result[j]

    return result


def bubble_sort_optimized(arr: List[T]) -> List[T]:
    """
    Optimized bubble sort with early termination.

    This version includes a flag to detect if any swaps were made during a pass.
    If no swaps occur, the array is already sorted and we can terminate early.

    Args:
        arr: List to be sorted

    Returns:
        New sorted list

    Time Complexity:
        - Best Case: O(n) when already sorted
        - Average/Worst: O(n²)
    Space Complexity: O(n) for the new list
    """
    if len(arr) <= 1:
        return arr.copy()

    result = arr.copy()
    n = len(result)

    for i in range(n):
        # Flag to optimize for already sorted arrays
        swapped = False

        for j in range(0, n - i - 1):
            if result[j] > result[j + 1]:
                result[j], result[j + 1] = result[j + 1], result[j]
                swapped = True

        # If no swaps occurred, array is sorted
        if not swapped:
            break

    return result


def bubble_sort_in_place(arr: List[T]) -> None:
    """
    In-place bubble sort implementation (optimized).

    Sorts the array in-place without creating a new list,
    minimizing space complexity.

    Args:
        arr: List to be sorted in-place

    Time Complexity: O(n) best case, O(n²) average/worst
    Space Complexity: O(1)
    """
    n = len(arr)

    for i in range(n):
        swapped = False

        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True

        if not swapped:
            break


def bubble_sort_recursive(arr: List[T], n: int = None) -> List[T]:
    """
    Recursive bubble sort implementation.

    Each recursive call performs one pass through the array,
    bubbling the largest element to the end.

    Args:
        arr: List to be sorted
        n: Size of the array portion to sort (used in recursion)

    Returns:
        New sorted list

    Time Complexity: O(n²)
    Space Complexity: O(n) for recursion stack + O(n) for new list
    """
    if n is None:
        result = arr.copy()
        return bubble_sort_recursive(result, len(result))

    # Base case: single element or empty
    if n <= 1:
        return arr

    # One pass of bubble sort
    # After this pass, the largest element will be at the end
    for i in range(n - 1):
        if arr[i] > arr[i + 1]:
            arr[i], arr[i + 1] = arr[i + 1], arr[i]

    # Recursively sort the first n-1 elements
    return bubble_sort_recursive(arr, n - 1)


def bubble_sort_functional(arr: List[T]) -> List[T]:
    """
    Functional-style bubble sort implementation.

    This implementation creates new lists for each iteration,
    following functional programming principles (immutability).

    Args:
        arr: List to be sorted

    Returns:
        New sorted list

    Time Complexity: O(n²)
    Space Complexity: O(n²) due to list copies
    """
    if len(arr) <= 1:
        return arr.copy()

    def single_pass(lst: List[T]) -> Tuple[List[T], bool]:
        """Perform one pass and return the result with swap status."""
        result = []
        swapped = False

        for i in range(len(lst) - 1):
            if lst[i] > lst[i + 1]:
                result.append(lst[i + 1])
                result.append(lst[i])
                swapped = True
                # Skip next element since we already added it
                if i + 2 < len(lst):
                    result.extend(lst[i + 2:])
                break
            else:
                result.append(lst[i])
        else:
            # Add the last element if we completed the loop
            result.append(lst[-1])

        return result, swapped

    # Simpler functional approach using list comprehension
    result = arr.copy()
    n = len(result)

    for i in range(n):
        swapped = False
        new_result = []

        for j in range(len(result) - 1):
            if result[j] > result[j + 1]:
                new_result.append(result[j + 1])
                new_result.append(result[j])
                result[j], result[j + 1] = result[j + 1], result[j]
                swapped = True

        if not swapped:
            break

    return result


def bubble_sort_with_comparator(arr: List[T], compare: Callable[[T, T], int] = None) -> List[T]:
    """
    Bubble sort with custom comparison function.

    Args:
        arr: List to be sorted
        compare: Comparison function (a, b) -> int
                 Returns: negative if a < b, 0 if a == b, positive if a > b

    Returns:
        New sorted list

    Example:
        # Sort in descending order
        bubble_sort_with_comparator([3, 1, 4], lambda a, b: b - a)
    """
    if compare is None:
        compare = lambda a, b: (a > b) - (a < b)

    result = arr.copy()
    n = len(result)

    for i in range(n):
        swapped = False

        for j in range(0, n - i - 1):
            if compare(result[j], result[j + 1]) > 0:
                result[j], result[j + 1] = result[j + 1], result[j]
                swapped = True

        if not swapped:
            break

    return result


def cocktail_sort(arr: List[T]) -> List[T]:
    """
    Cocktail Shaker Sort (bidirectional bubble sort).

    An optimized version of bubble sort that sorts in both directions
    alternately, which can be more efficient for certain data patterns.

    Args:
        arr: List to be sorted

    Returns:
        New sorted list

    Time Complexity: O(n²) worst case, but often faster than standard bubble sort
    Space Complexity: O(n)
    """
    if len(arr) <= 1:
        return arr.copy()

    result = arr.copy()
    start = 0
    end = len(result) - 1
    swapped = True

    while swapped:
        swapped = False

        # Forward pass (like bubble sort)
        for i in range(start, end):
            if result[i] > result[i + 1]:
                result[i], result[i + 1] = result[i + 1], result[i]
                swapped = True

        if not swapped:
            break

        swapped = False
        end -= 1

        # Backward pass
        for i in range(end - 1, start - 1, -1):
            if result[i] > result[i + 1]:
                result[i], result[i + 1] = result[i + 1], result[i]
                swapped = True

        start += 1

    return result


def is_sorted(arr: List[T]) -> bool:
    """Check if array is sorted in ascending order."""
    for i in range(len(arr) - 1):
        if arr[i] > arr[i + 1]:
            return False
    return True


def demonstrate_bubble_sort():
    """Demonstrate various bubble sort implementations."""
    print("🫧 Bubble Sort Implementation in Python")
    print("=" * 50)

    # Test data - comprehensive edge cases
    test_arrays = [
        ([64, 34, 25, 12, 22, 11, 90], "Random array"),
        ([5, 2, 8, 6, 1, 9, 4], "Small random array"),
        ([1], "Single element"),
        ([], "Empty array"),
        ([3, 3, 3, 3, 3], "All duplicates"),
        ([9, 8, 7, 6, 5, 4, 3, 2, 1], "Reverse sorted"),
        ([1, 2, 3, 4, 5], "Already sorted"),
        ([5, 1, 4, 2, 3], "Nearly sorted"),
    ]

    print("\n📋 Basic Sorting Tests:")
    print("-" * 50)

    for arr, description in test_arrays:
        original = arr.copy()

        # Test different implementations
        iterative_result = bubble_sort_iterative(arr)
        optimized_result = bubble_sort_optimized(arr)
        recursive_result = bubble_sort_recursive(arr)
        cocktail_result = cocktail_sort(arr)

        # Test in-place
        inplace_result = arr.copy()
        bubble_sort_in_place(inplace_result)

        print(f"\nTest: {description}")
        print(f"Original:  {original}")
        print(f"Sorted:    {iterative_result}")

        # Verify all results are correct and equal
        results = [iterative_result, optimized_result, recursive_result,
                   cocktail_result, inplace_result]
        all_correct = all(is_sorted(result) for result in results)
        all_equal = all(result == iterative_result for result in results)

        status = "✓" if all_correct and all_equal else "✗"
        print(f"All implementations match: {status}")

    print("\n" + "-" * 50)

    # String sorting
    words = ["banana", "apple", "cherry", "date", "elderberry"]
    sorted_words = bubble_sort_optimized(words)

    print(f"\n🔤 Word sorting:")
    print(f"Original:     {words}")
    print(f"Alphabetical: {sorted_words}")

    # Custom comparison
    numbers = [3, 1, 4, 1, 5, 9, 2, 6]
    desc_sorted = bubble_sort_with_comparator(numbers, lambda a, b: b - a)

    print(f"\n🔢 Custom comparison (descending):")
    print(f"Original:   {numbers}")
    print(f"Descending: {desc_sorted}")


def performance_benchmark():
    """Benchmark different bubble sort implementations."""
    print(f"\n\n⚡ Performance Benchmark")
    print("=" * 70)

    sizes = [100, 500, 1000, 2000]

    # Test different data patterns
    patterns = {
        "Random": lambda n: [random.randint(1, 1000) for _ in range(n)],
        "Sorted": lambda n: list(range(n)),
        "Reversed": lambda n: list(range(n, 0, -1)),
        "Nearly Sorted": lambda n: list(range(n)) if n < 10 else
                         list(range(n-5)) + [n-1, n-3, n-2, n-4, n-5] + list(range(n-5, n)),
    }

    methods = {
        "Iterative": bubble_sort_iterative,
        "Optimized": bubble_sort_optimized,
        "Recursive": bubble_sort_recursive,
        "Cocktail": cocktail_sort,
        "Built-in": lambda arr: sorted(arr),
    }

    for pattern_name, pattern_gen in patterns.items():
        print(f"\n{pattern_name} Data:")
        print("Size".rjust(8), end="")
        for method in methods:
            print(method.rjust(12), end="")
        print()
        print("-" * (8 + 12 * len(methods)))

        for size in sizes:
            # Generate test data
            test_data = pattern_gen(size)

            print(f"{size:8d}", end="")

            for method_name, method_func in methods.items():
                # Skip recursive for large sizes (stack overflow risk)
                if method_name == "Recursive" and size > 1000:
                    print(f"{'N/A':>12}", end="")
                    continue

                # Benchmark
                try:
                    start_time = time.perf_counter()
                    result = method_func(test_data.copy())
                    end_time = time.perf_counter()

                    elapsed_ms = (end_time - start_time) * 1000
                    print(f"{elapsed_ms:11.2f}ms", end="")

                    # Verify correctness
                    if not is_sorted(result if result else []):
                        print(" ✗", end="")
                except RecursionError:
                    print(f"{'OVERFLOW':>12}", end="")

            print()


def count_comparisons_and_swaps(arr: List[T]) -> Tuple[int, int]:
    """
    Count the number of comparisons and swaps made during bubble sort.

    Args:
        arr: List to be sorted

    Returns:
        Tuple of (comparisons, swaps)
    """
    result = arr.copy()
    n = len(result)
    comparisons = 0
    swaps = 0

    for i in range(n):
        for j in range(0, n - i - 1):
            comparisons += 1
            if result[j] > result[j + 1]:
                result[j], result[j + 1] = result[j + 1], result[j]
                swaps += 1

    return comparisons, swaps


def analyze_algorithm():
    """Analyze bubble sort behavior with different inputs."""
    print(f"\n\n🔍 Algorithm Analysis")
    print("=" * 50)

    test_cases = [
        ([5, 2, 8, 6, 1], "Random"),
        ([1, 2, 3, 4, 5], "Already sorted"),
        ([5, 4, 3, 2, 1], "Reverse sorted"),
    ]

    for arr, description in test_cases:
        comparisons, swaps = count_comparisons_and_swaps(arr)
        n = len(arr)

        print(f"\n{description}: {arr}")
        print(f"Array size (n): {n}")
        print(f"Comparisons: {comparisons} (theoretical max: {n * (n - 1) // 2})")
        print(f"Swaps: {swaps}")
        print(f"Efficiency: {(1 - swaps / max(comparisons, 1)) * 100:.1f}% "
              f"(fewer swaps is better)")


class BubbleSorter:
    """
    Object-oriented wrapper for bubble sort with statistics tracking.
    """

    def __init__(self):
        self.comparisons = 0
        self.swaps = 0
        self.iterations = 0

    def sort(self, arr: List[T]) -> List[T]:
        """Sort array and track statistics."""
        self.reset_stats()
        result = arr.copy()
        n = len(result)

        for i in range(n):
            self.iterations += 1
            swapped = False

            for j in range(0, n - i - 1):
                self.comparisons += 1
                if result[j] > result[j + 1]:
                    result[j], result[j + 1] = result[j + 1], result[j]
                    self.swaps += 1
                    swapped = True

            if not swapped:
                break

        return result

    def reset_stats(self):
        """Reset statistics counters."""
        self.comparisons = 0
        self.swaps = 0
        self.iterations = 0

    def get_stats(self) -> dict:
        """Get sorting statistics."""
        return {
            "comparisons": self.comparisons,
            "swaps": self.swaps,
            "iterations": self.iterations,
        }


if __name__ == "__main__":
    demonstrate_bubble_sort()
    performance_benchmark()
    analyze_algorithm()

    # Demonstrate OOP approach
    print(f"\n\n📊 Object-Oriented Approach")
    print("=" * 50)

    sorter = BubbleSorter()
    test_array = [64, 34, 25, 12, 22, 11, 90]

    sorted_array = sorter.sort(test_array)
    stats = sorter.get_stats()

    print(f"Original: {test_array}")
    print(f"Sorted:   {sorted_array}")
    print(f"Statistics: {stats}")

    print("\n✨ Bubble Sort demonstration complete!")

File: sorting/bubblesort.js
/**
 * Bubble Sort Algorithm Implementation in JavaScript
 *
 * Time Complexity:
 * - Best Case: O(n) - when array is already sorted (optimized version)
 * - Average Case: O(n²)
 * - Worst Case: O(n²) - when array is reverse sorted
 * Space Complexity: O(1) for in-place, O(n) for functional approach
 *
 * Bubble Sort works by repeatedly stepping through the list, comparing adjacent
 * elements and swapping them if they are in the wrong order. The pass through
 * the list is repeated until the list is sorted.
 *
 * JavaScript features:
 * - ES6+ syntax (arrow functions, destructuring, spread operator)
 * - Array extension methods
 * - Functional programming patterns
 * - Performance measurement using performance.now()
 */

/**
 * Basic iterative bubble sort implementation.
 *
 * This is the standard bubble sort algorithm that compares and swaps
 * adjacent elements until the entire array is sorted.
 *
 * @template T
 * @param {T[]} arr - Array to be sorted
 * @returns {T[]} New sorted array
 *
 * Time Complexity: O(n²) in all cases (no optimization)
 * Space Complexity: O(n) for the new array
 */
function bubbleSortIterative(arr) {
    if (arr.length <= 1) {
        return [...arr];
    }

    const result = [...arr];
    const n = result.length;

    // Outer loop for number of passes
    for (let i = 0; i < n; i++) {
        // Inner loop for comparisons
        // After each pass, the largest element "bubbles up" to its position
        for (let j = 0; j < n - i - 1; j++) {
            if (result[j] > result[j + 1]) {
                // Swap adjacent elements using destructuring
                [result[j], result[j + 1]] = [result[j + 1], result[j]];
            }
        }
    }

    return result;
}

/**
 * Optimized bubble sort with early termination.
 *
 * This version includes a flag to detect if any swaps were made during a pass.
 * If no swaps occur, the array is already sorted and we can terminate early.
 *
 * @template T
 * @param {T[]} arr - Array to be sorted
 * @returns {T[]} New sorted array
 *
 * Time Complexity:
 *   - Best Case: O(n) when already sorted
 *   - Average/Worst: O(n²)
 * Space Complexity: O(n) for the new array
 */
function bubbleSortOptimized(arr) {
    if (arr.length <= 1) {
        return [...arr];
    }

    const result = [...arr];
    const n = result.length;

    for (let i = 0; i < n; i++) {
        // Flag to optimize for already sorted arrays
        let swapped = false;

        for (let j = 0; j < n - i - 1; j++) {
            if (result[j] > result[j + 1]) {
                [result[j], result[j + 1]] = [result[j + 1], result[j]];
                swapped = true;
            }
        }

        // If no swaps occurred, array is sorted
        if (!swapped) {
            break;
        }
    }

    return result;
}

/**
 * In-place bubble sort implementation (optimized).
 *
 * Sorts the array in-place without creating a new array,
 * minimizing space complexity.
 *
 * @template T
 * @param {T[]} arr - Array to be sorted in-place
 * @returns {void}
 *
 * Time Complexity: O(n) best case, O(n²) average/worst
 * Space Complexity: O(1)
 */
function bubbleSortInPlace(arr) {
    const n = arr.length;

    for (let i = 0; i < n; i++) {
        let swapped = false;

        for (let j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
                swapped = true;
            }
        }

        if (!swapped) {
            break;
        }
    }
}

/**
 * Recursive bubble sort implementation.
 *
 * Each recursive call performs one pass through the array,
 * bubbling the largest element to the end.
 *
 * @template T
 * @param {T[]} arr - Array to be sorted
 * @param {number} [n=arr.length] - Size of the array portion to sort
 * @returns {T[]} The same array (sorted in-place for recursion efficiency)
 *
 * Time Complexity: O(n²)
 * Space Complexity: O(n) for recursion stack
 */
function bubbleSortRecursive(arr, n = null) {
    // Initialize on first call
    if (n === null) {
        const result = [...arr];
        return bubbleSortRecursive(result, result.length);
    }

    // Base case: single element or empty
    if (n <= 1) {
        return arr;
    }

    // One pass of bubble sort
    // After this pass, the largest element will be at the end
    for (let i = 0; i < n - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            [arr[i], arr[i + 1]] = [arr[i + 1], arr[i]];
        }
    }

    // Recursively sort the first n-1 elements
    return bubbleSortRecursive(arr, n - 1);
}

/**
 * Functional-style bubble sort implementation.
 *
 * This implementation uses functional programming principles,
 * creating new arrays for immutability.
 *
 * @template T
 * @param {T[]} arr - Array to be sorted
 * @returns {T[]} New sorted array
 *
 * Time Complexity: O(n²)
 * Space Complexity: O(n)
 */
function bubbleSortFunctional(arr) {
    if (arr.length <= 1) {
        return [...arr];
    }

    const performPass = (array) => {
        let swapped = false;
        const newArray = [...array];

        for (let i = 0; i < newArray.length - 1; i++) {
            if (newArray[i] > newArray[i + 1]) {
                [newArray[i], newArray[i + 1]] = [newArray[i + 1], newArray[i]];
                swapped = true;
            }
        }

        return { array: newArray, swapped };
    };

    let result = [...arr];
    let continueSort = true;

    while (continueSort) {
        const { array, swapped } = performPass(result);
        result = array;
        continueSort = swapped;
    }

    return result;
}

/**
 * Bubble sort with custom comparison function.
 *
 * @template T
 * @param {T[]} arr - Array to be sorted
 * @param {(a: T, b: T) => number} [compareFn] - Comparison function
 * @returns {T[]} New sorted array
 *
 * Example:
 *   bubbleSortWithComparator([3, 1, 4], (a, b) => b - a) // Descending order
 */
function bubbleSortWithComparator(arr, compareFn = (a, b) => a > b ? 1 : a < b ? -1 : 0) {
    if (arr.length <= 1) {
        return [...arr];
    }

    const result = [...arr];
    const n = result.length;

    for (let i = 0; i < n; i++) {
        let swapped = false;

        for (let j = 0; j < n - i - 1; j++) {
            if (compareFn(result[j], result[j + 1]) > 0) {
                [result[j], result[j + 1]] = [result[j + 1], result[j]];
                swapped = true;
            }
        }

        if (!swapped) {
            break;
        }
    }

    return result;
}

/**
 * Cocktail Shaker Sort (bidirectional bubble sort).
 *
 * An optimized version of bubble sort that sorts in both directions
 * alternately, which can be more efficient for certain data patterns.
 *
 * @template T
 * @param {T[]} arr - Array to be sorted
 * @returns {T[]} New sorted array
 *
 * Time Complexity: O(n²) worst case, but often faster than standard bubble sort
 * Space Complexity: O(n)
 */
function cocktailSort(arr) {
    if (arr.length <= 1) {
        return [...arr];
    }

    const result = [...arr];
    let start = 0;
    let end = result.length - 1;
    let swapped = true;

    while (swapped) {
        swapped = false;

        // Forward pass (like bubble sort)
        for (let i = start; i < end; i++) {
            if (result[i] > result[i + 1]) {
                [result[i], result[i + 1]] = [result[i + 1], result[i]];
                swapped = true;
            }
        }

        if (!swapped) {
            break;
        }

        swapped = false;
        end--;

        // Backward pass
        for (let i = end - 1; i >= start; i--) {
            if (result[i] > result[i + 1]) {
                [result[i], result[i + 1]] = [result[i + 1], result[i]];
                swapped = true;
            }
        }

        start++;
    }

    return result;
}

/**
 * Check if array is sorted in ascending order.
 *
 * @template T
 * @param {T[]} arr - Array to check
 * @returns {boolean} True if sorted
 */
function isSorted(arr) {
    for (let i = 0; i < arr.length - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

/**
 * Count comparisons and swaps during bubble sort.
 *
 * @template T
 * @param {T[]} arr - Array to analyze
 * @returns {{ comparisons: number, swaps: number }}
 */
function countComparisonsAndSwaps(arr) {
    const result = [...arr];
    const n = result.length;
    let comparisons = 0;
    let swaps = 0;

    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n - i - 1; j++) {
            comparisons++;
            if (result[j] > result[j + 1]) {
                [result[j], result[j + 1]] = [result[j + 1], result[j]];
                swaps++;
            }
        }
    }

    return { comparisons, swaps };
}

/**
 * Array extension methods for convenient sorting.
 */
if (typeof Array.prototype.bubbleSort === 'undefined') {
    /**
     * Extension method to sort array using bubble sort.
     * @returns {Array} New sorted array
     */
    Array.prototype.bubbleSort = function() {
        return bubbleSortOptimized(this);
    };

    /**
     * Extension method to sort array in-place using bubble sort.
     * @returns {Array} The same array (sorted)
     */
    Array.prototype.bubbleSortInPlace = function() {
        bubbleSortInPlace(this);
        return this;
    };
}

/**
 * BubbleSorter class with statistics tracking.
 */
class BubbleSorter {
    constructor() {
        this.comparisons = 0;
        this.swaps = 0;
        this.iterations = 0;
    }

    /**
     * Sort array and track statistics.
     * @template T
     * @param {T[]} arr - Array to sort
     * @returns {T[]} Sorted array
     */
    sort(arr) {
        this.resetStats();
        const result = [...arr];
        const n = result.length;

        for (let i = 0; i < n; i++) {
            this.iterations++;
            let swapped = false;

            for (let j = 0; j < n - i - 1; j++) {
                this.comparisons++;
                if (result[j] > result[j + 1]) {
                    [result[j], result[j + 1]] = [result[j + 1], result[j]];
                    this.swaps++;
                    swapped = true;
                }
            }

            if (!swapped) {
                break;
            }
        }

        return result;
    }

    resetStats() {
        this.comparisons = 0;
        this.swaps = 0;
        this.iterations = 0;
    }

    getStats() {
        return {
            comparisons: this.comparisons,
            swaps: this.swaps,
            iterations: this.iterations
        };
    }
}

/**
 * Demonstrate various bubble sort implementations.
 */
function demonstrateBubbleSort() {
    console.log("🫧 Bubble Sort Implementation in JavaScript");
    console.log("=".repeat(50));

    // Test data - comprehensive edge cases
    const testArrays = [
        { arr: [64, 34, 25, 12, 22, 11, 90], desc: "Random array" },
        { arr: [5, 2, 8, 6, 1, 9, 4], desc: "Small random array" },
        { arr: [1], desc: "Single element" },
        { arr: [], desc: "Empty array" },
        { arr: [3, 3, 3, 3, 3], desc: "All duplicates" },
        { arr: [9, 8, 7, 6, 5, 4, 3, 2, 1], desc: "Reverse sorted" },
        { arr: [1, 2, 3, 4, 5], desc: "Already sorted" },
        { arr: [5, 1, 4, 2, 3], desc: "Nearly sorted" }
    ];

    console.log("\n📋 Basic Sorting Tests:");
    console.log("-".repeat(50));

    testArrays.forEach(({ arr, desc }) => {
        const original = [...arr];

        // Test different implementations
        const iterativeResult = bubbleSortIterative(arr);
        const optimizedResult = bubbleSortOptimized(arr);
        const recursiveResult = bubbleSortRecursive(arr);
        const cocktailResult = cocktailSort(arr);

        // Test in-place
        const inplaceResult = [...arr];
        bubbleSortInPlace(inplaceResult);

        console.log(`\nTest: ${desc}`);
        console.log(`Original:  [${original.join(', ')}]`);
        console.log(`Sorted:    [${iterativeResult.join(', ')}]`);

        // Verify all results are correct and equal
        const results = [iterativeResult, optimizedResult, recursiveResult,
                        cocktailResult, inplaceResult];
        const allCorrect = results.every(result => isSorted(result));
        const allEqual = results.every(result =>
            JSON.stringify(result) === JSON.stringify(iterativeResult));

        const status = allCorrect && allEqual ? "✓" : "✗";
        console.log(`All implementations match: ${status}`);
    });

    console.log("\n" + "-".repeat(50));

    // String sorting
    const words = ["banana", "apple", "cherry", "date", "elderberry"];
    const sortedWords = bubbleSortOptimized(words);

    console.log("\n🔤 Word sorting:");
    console.log(`Original:     [${words.join(', ')}]`);
    console.log(`Alphabetical: [${sortedWords.join(', ')}]`);

    // Custom comparison
    const numbers = [3, 1, 4, 1, 5, 9, 2, 6];
    const descSorted = bubbleSortWithComparator(numbers, (a, b) => b - a);

    console.log("\n🔢 Custom comparison (descending):");
    console.log(`Original:   [${numbers.join(', ')}]`);
    console.log(`Descending: [${descSorted.join(', ')}]`);

    // Array extension methods
    console.log("\n🔧 Extension Methods:");
    const testExt = [5, 2, 8, 1, 9];
    console.log(`Original: [${testExt.join(', ')}]`);
    console.log(`Using .bubbleSort(): [${testExt.bubbleSort().join(', ')}]`);
}

/**
 * Benchmark different bubble sort implementations.
 */
function performanceBenchmark() {
    console.log("\n\n⚡ Performance Benchmark");
    console.log("=".repeat(70));

    const sizes = [100, 500, 1000, 2000];

    // Test different data patterns
    const patterns = {
        "Random": (n) => Array.from({ length: n }, () => Math.floor(Math.random() * 1000)),
        "Sorted": (n) => Array.from({ length: n }, (_, i) => i),
        "Reversed": (n) => Array.from({ length: n }, (_, i) => n - i),
        "Nearly Sorted": (n) => {
            const arr = Array.from({ length: n }, (_, i) => i);
            if (n >= 10) {
                // Swap a few elements
                for (let i = 0; i < Math.min(5, n / 10); i++) {
                    const idx = Math.floor(Math.random() * n);
                    const idx2 = Math.floor(Math.random() * n);
                    [arr[idx], arr[idx2]] = [arr[idx2], arr[idx]];
                }
            }
            return arr;
        }
    };

    const methods = {
        "Iterative": bubbleSortIterative,
        "Optimized": bubbleSortOptimized,
        "Recursive": bubbleSortRecursive,
        "Cocktail": cocktailSort,
        "Built-in": (arr) => [...arr].sort((a, b) => a - b)
    };

    for (const [patternName, patternGen] of Object.entries(patterns)) {
        console.log(`\n${patternName} Data:`);

        // Header
        let header = "Size".padEnd(8);
        for (const method of Object.keys(methods)) {
            header += method.padStart(12);
        }
        console.log(header);
        console.log("-".repeat(8 + 12 * Object.keys(methods).length));

        for (const size of sizes) {
            const testData = patternGen(size);
            let row = size.toString().padEnd(8);

            for (const [methodName, methodFunc] of Object.entries(methods)) {
                // Skip recursive for large sizes
                if (methodName === "Recursive" && size > 1000) {
                    row += "N/A".padStart(12);
                    continue;
                }

                try {
                    const start = performance.now();
                    const result = methodFunc([...testData]);
                    const end = performance.now();

                    const elapsedMs = end - start;
                    row += `${elapsedMs.toFixed(2)}ms`.padStart(12);

                    // Verify correctness
                    if (!isSorted(result)) {
                        row += " ✗";
                    }
                } catch (error) {
                    row += "ERROR".padStart(12);
                }
            }

            console.log(row);
        }
    }
}

/**
 * Analyze bubble sort behavior with different inputs.
 */
function analyzeAlgorithm() {
    console.log("\n\n🔍 Algorithm Analysis");
    console.log("=".repeat(50));

    const testCases = [
        { arr: [5, 2, 8, 6, 1], desc: "Random" },
        { arr: [1, 2, 3, 4, 5], desc: "Already sorted" },
        { arr: [5, 4, 3, 2, 1], desc: "Reverse sorted" }
    ];

    testCases.forEach(({ arr, desc }) => {
        const { comparisons, swaps } = countComparisonsAndSwaps(arr);
        const n = arr.length;
        const theoreticalMax = n * (n - 1) / 2;

        console.log(`\n${desc}: [${arr.join(', ')}]`);
        console.log(`Array size (n): ${n}`);
        console.log(`Comparisons: ${comparisons} (theoretical max: ${theoreticalMax})`);
        console.log(`Swaps: ${swaps}`);
        console.log(`Efficiency: ${((1 - swaps / Math.max(comparisons, 1)) * 100).toFixed(1)}% (fewer swaps is better)`);
    });
}

// Main execution
if (typeof require !== 'undefined' && require.main === module) {
    demonstrateBubbleSort();
    performanceBenchmark();
    analyzeAlgorithm();

    // Demonstrate OOP approach
    console.log("\n\n📊 Object-Oriented Approach");
    console.log("=".repeat(50));

    const sorter = new BubbleSorter();
    const testArray = [64, 34, 25, 12, 22, 11, 90];

    const sortedArray = sorter.sort(testArray);
    const stats = sorter.getStats();

    console.log(`Original: [${testArray.join(', ')}]`);
    console.log(`Sorted:   [${sortedArray.join(', ')}]`);
    console.log(`Statistics:`, stats);

    console.log("\n✨ Bubble Sort demonstration complete!");
}

// Export for use as module
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        bubbleSortIterative,
        bubbleSortOptimized,
        bubbleSortInPlace,
        bubbleSortRecursive,
        bubbleSortFunctional,
        bubbleSortWithComparator,
        cocktailSort,
        isSorted,
        BubbleSorter,
        demonstrateBubbleSort,
        performanceBenchmark,
        analyzeAlgorithm
    };
}
